src/main/scala/is/hail/expr/ir/Recur.scala

```diff
@@ -0,0 +1,42 @@
+package is.hail.expr.ir
+
+import is.hail.annotations._
+import is.hail.asm4s._
+import is.hail.expr
+import is.hail.expr.{Type, Aggregator, TAggregable, TArray, TBoolean, TContainer, TFloat32, TFloat64, TInt32, TInt64, TStruct}
+import is.hail.utils._
+import org.objectweb.asm.Opcodes._
+import org.objectweb.asm.tree._
+
+object Recur {
+  def apply(f: IR => IR)(ir: IR): IR = ir match {
+    case I32(x) => ir
+    case I64(x) => ir
+    case F32(x) => ir
+    case F64(x) => ir
+    case True() => ir
+    case False() => ir
+    case Cast(v, typ) => Cast(f(v), typ)
+    case NA(typ) => ir
+    case MapNA(name, value, body, typ) => MapNA(name, f(value), f(body), typ)
+    case IsNA(value) => IsNA(value)
+    case If(cond, cnsq, altr, typ) => If(f(cond), f(cnsq), f(altr), typ)
+    case Let(name, value, body, typ) => Let(name, f(value), f(body), typ)
+    case Ref(name, typ) => ir
+    case ApplyBinaryPrimOp(op, l, r, typ) => ApplyBinaryPrimOp(op, f(l), f(r), typ)
+    case ApplyUnaryPrimOp(op, x, typ) => ApplyUnaryPrimOp(op, f(x), typ)
+    case MakeArray(args, typ) => MakeArray(args map f, typ)
+    case MakeArrayN(len, elementType) => MakeArrayN(f(len), elementType)
+    case ArrayRef(a, i, typ) => ArrayRef(f(a), f(i), typ)
+    case ArrayMissingnessRef(a, i) => ArrayMissingnessRef(f(a), i)
```

Comment:
`f(i)`?

Reply:
no, which input you're referencing must be known at expr-language-compile-time

Reply:
Then I'm very confused. `i` has type `IR`. If you're sure it shouldn't be `f(i)`, then can you explain it to me in person?

Reply:
Oh whoops, I somehow read this as `InMissingness` (and apparently did the same thing when implementing this). You're correct.